timeline-chart/src/layer/time-graph-chart.ts

```diff
@@ -29,13 +30,18 @@ export const keyBoardNavs: Record<string, Array<string>> = {
     "zoomin": ['w', 'i'],
     "zoomout": ['s', 'k'],
     "panleft": ['a', 'j'],
-    "panright": ['d', 'l']
+    "panright": ['d', 'l'],
+    "undo": ['z'],
+    "redo": ['y'],
 }
 
 export type TimeGraphRowStyleHook = (row: TimelineChart.TimeGraphRowModel) => TimeGraphRowStyle | undefined;
 
 export class TimeGraphChart extends TimeGraphChartLayer {
 
+    protected undoStack: bigint[][];
+    protected redoStack: bigint[][];
```

Comment:
I tried to use stack to hold function such as adjustZoom or moveHorizontally, which causes bugs. For example, if we store moveHorizontally(-5) in the redoStack and do zoom in, the zoom factor would be different. As a result, to redo moveHorizontally(-5) will leads to a different result instead of the original state.